libraries/ESP8266WiFi/src/WiFiServer.cpp

```diff
@@ -77,13 +82,21 @@ void WiFiServer::begin(uint16_t port) {
         return;
     }
 
+#if LWIP_VERSION_MAJOR == 1
+    (void)backlog;
     tcp_pcb* listen_pcb = tcp_listen(pcb);
+#else
+    if (backlog == 0)
```

Comment:
instead of mapping 0=>5, how about malling 0=>disabled backlog?

Reply:
Good question ! (disabling it by default is like setting to max `0xff`)
The goal of the feature is to protect the esp from OOM at worst or heap fragmentation at least when connection requests rate is greater than or around service time rate. My answer would be another question: what would we like by default for the average arduino sketch ?

Reply:
I think you had it right before: the default should likely be the define, 5. My question here is about disabling it. 
Option 1: 0 means disabled. You could have a special case to go through a different code path and not do backlog at all. 
Option 2: 0 means max possible. If the arg is 0, replace the value with the max possible backlog number 0xff.
Option 3: 0 as arg is not allowed. A define or constexpr is provided so that the user can set it to th3 max value possible (0xff).

I prefer option 1, but will settle for option 2. I would prefer to avoid option 3.

Reply:
I have a sub-question.  This is about desired behavior when multiple clients access the unit, but what if a connection is not closed by a client?
So this brings to (my) mind the question whether a single counter here is enough?
This may lead to another and bigger change, so my expectation (desire?) is that it will NOT be part of this change, but it may give some ideas for this counter.

Is the WiFi Server, or LWIP2 aware of the continuation of receiving packets by the client?
If so, then there may be a difference in desired behavior when all data was sent but the client simply didn't close the connection, compared to being in the middle of a transfer.
There is also the other way around, when the client deliberately is not waiting for a reply/ack, is the connection then still considered to be an existing one?

So maybe we should accept a (single?) new one, even though the set amount is in use.

Reply:
In the current state of this PR, the possible number of connections is limited to the max (5 by default).
We can change this behaviour that way:
- we can have an unlimited number of *accepted* connections:
  that would answer to you concern I think,
- ***and*** keep number of *pending* connections limited to 5:
  - if user doesn't care at all and treat one connection at a time (timeouts are meant to close dying clients), then ram usage is protected
  - if user wants to be able to handle many (>>5) connections simultaneously, it is only needed to **stop** checking `WiFiServer::available()` on user constraints.
    From RAM pov, the worst situation would be the already accepted connections, 5 connections in ram waiting to be acknowledged with `WiFiServer.available()` and the other newer requests waiting&retrying outside, not using a single byte of ram.

I'll make some tests and commit. Thanks for your sub-question @TD-er . Until now I was wondering why lwIP's `tcp_backlog_accepted()` was not included in `tcp_close()`. Their documentation on this is very poor. Now with your question, I think I understand why: We can call it before `tcp_close()`, once a connection is accepted.

Reply:
> we can have an unlimited number of accepted connections:
> that would answer to you concern I think,

Well, how many resources do "accepted connections" take?
And how do clients react on "accepted-but-never-answered" connections?
Will they actively close them after some timeout? If not, then the ESP may try to serve to a non existing connection

Reply:
The issue is about the number of silently accepted connection.

Without backlog feature, there is no limitation and ESP can OOM because of many external requests connecting to a server port. Once connected (this is automatic, internal to lwIP, with a callback to store them in a waiting list) and even when sketch does not `WiFiServer::available()` them, ram is allocated (=> oom) while clients patiently wait for an answer. No timeout in that case (except when clients want to).

With an example: if one starts many simultaneous `telnet <esp-ip> <port>`, a `tcp_pcb` is allocated in esp for every single request. In that case OOM can happen even when `WiFiServer::available()` is called but requests keep coming faster than served and closed.

> Well, how many resources do "accepted connections" take?

some tens of bytes (`sizeof(tcp_pcb)`)

> And how do clients react on "accepted-but-never-answered" connections?
> Will they actively close them after some timeout?

Application defined. `telnet` waits forever.

> If not, then the ESP may try to serve to a non existing connection

There are timeouts on application level (our webserver is 5secs), and also at tcp levels. The latter can be very long (hours) before a connection is closed (that's why there is the tcp-keepalive option, to get a shorter/configurable internal tcp timeout).
On the other hand, if the client closes the connection, ram is released on esp side.

There is also another software way to deal with too many incoming connections which has not been addressed so far because it is not clean:
When the internal lwIP callback is called but the already accepted list is too large, WiFiServer could close the connection. That's bad behaviour. Backlog limitation handles this differently: it does not answer immediately until resources allow to answer properly.

Reply:
@devyte 
As discussed, 0 for backlog is now a forbidden value which will not open the server port.